### PR TITLE
build: avoid repetitions when enabling warnings in configure.ac

### PR DESCRIPTION
**Pull request by vasild**
Introduce a macro `APPEND_FLAG_IF_AVAILABLE()` and use that.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Concept ACK b26f8a0bb9bc6bb1066291e6752b7e33bea89cf9

**Comment by brakmic** [comment]
Concept ACK b26f8a0bb9bc6bb1066291e6752b7e33bea89cf9

**Comment by practicalswift** [comment]
Concept ACK

Very nice initiative: this repetition has annoyed me :)

File: configure.ac [hunk 1]
```diff
@@ -333,31 +333,36 @@ if test x$use_sanitizers != x; then
     ]],[[]])])
 fi
 
+dnl APPEND_FLAG_IF_AVAILABLE([APPEND_TO_THIS_VARIABLE], [-WallExtra])
+AC_DEFUN([APPEND_FLAG_IF_AVAILABLE], [
```
Comment (promag) [inline configure.ac:337]: Could also define `SET_FLAG_IF_AVAILABLE` and use throughout.

**Review by promag — COMMENTED** [review]
Concept ACK.

File: configure.ac (hunk 1 above)
Comment (vasild) [reply configure.ac:337]: That would be a little bit less straight forward. While such code would definitely benefit from it:
```
AX_CHECK_COMPILE_FLAG([-msse4.2],[[SSE42_CXXFLAGS="-msse4.2"]],,[[$CXXFLAG_WERROR]])
```
and become:
```
SET_FLAG_IF_AVAILABLE([SSE42_CXXFLAGS], [-msse4.2])
```
there are other usages where the code calls `AC_MSG_ERROR()` if the flag is not available or does not compile with an additional `$CXXFLAG_WERROR`.

I will consider this as a separate PR.

Comment (promag) [reply configure.ac:337]: Yeah didn't check if all follow the same pattern. Different PR is fine.

**Comment by fanquake** [comment]
Concept ~0. Saving same verbosity is nice, but the downsides are now having a macro that obscures what's happening, and is applied inconsistently throughout configure; less than half of our `AX_CHECK_COMPILE_FLAG` calls are replaced with it.

This approach would also not work with the change currently proposed in #18882 (we'd end up with duplicate `-Wformat` flags). So let's decide how we want to fix the issue in that PR, before merging something like this.

**Comment by vasild** [comment]
Rebased.

**Comment by vasild** [comment]
Rebased.

**Comment by vasild** [comment]
Rebased.

**Comment by vasild** [comment]
Rebased.

**Comment by vasild** [comment]
Rebased.

**Comment by vasild** [comment]
Rebased.

**Comment by vasild** [comment]
Rebased.

**Comment by vasild** [comment]
Rebased due to conflicts.

@fanquake, what's your take on this, now that https://github.com/bitcoin/bitcoin/pull/18882 has been merged?

**Comment by fanquake** [comment]
> @fanquake, what's your take on this, now that #18882 has been merged?

My take is still the same as [my comment above](https://github.com/bitcoin/bitcoin/pull/18857#issuecomment-623902841). I'm not convinced introducing an inconsistently used macro, to essentially save some line length, is a huge win. In general I prefer that we are explicit/verbose when it comes to anything build system related.

**Comment by vasild** [comment]
Closing this because it needs frequent rebasing and it is not moving anywhere.

It has 4 "Concept ACK" and 1 "Concept ~0".

The point of this PR is to save repetitions, not line length. The introduced macro is not used everywhere (to replace `AX_CHECK_COMPILE_FLAG`), but on the places where it is used the usage is consistent.

If there is interest in this, I would be happy to reopen and rebase it.

**Comment by hebasto** [comment]
I tend to agree with @fanquake:
> In general I prefer that we are explicit/verbose when it comes to anything build system related.

**Pull request closed without merging**